### ISSUE TITLE
WASM vs WAT vs compiled languages

### ISSUE THREAD

**Issue by joshgoebel** [nesbox/TIC-80#1775]
Related #1259.

The original implementation had this idea of the editor being WAT and the runtime being WASM, but I think this is flawed... the `script` metadata being `WASM` doesn't imply WAT is the *source language*... now WAT is simply a textual format of WASM true, but the source language might have been C, C++, Go, Rust, or many others... so I think expecting `wasm` to be written in `wat` is already broken conceptually.  

I feel like if we wanted someone to use the built-in editor to write raw WAT (which doesn't seem like something many people would want), that the `script` would be `wat`... and that when you hit run the `wat` would be compiled to WASM and run "on the fly"... there might never be a BINARY chunk at all, least not one that was persisted.

So hence a `wat` project might potentially make sense, but again it wouldn't have a `BINARY` chunk... it would just be WAT code that was compiled as needed by TIC-80 to run the cartridge.  I don't see the need to store both since then you have no idea which is canonical (one has to be)... and once one is, why not just keep that?  Since the source can also has comments, etc... it seems to make sense for that to be canonical.

Of course it would technically be possible to "compile" a `wat` project into a `wasm` project (and remove the source, or leave it just as a reference), but I think I see the two are two different "script" meta-values...

---

Mostly I don't like the idea of mixing wasm/wat because it blurs the boundary with every other compiled language (and many we will NEVER include compilers for).  For example would we ever have a `script: rust` tag?  I don't think so, that's a huge ecosystem/toolchain and likely always going to be external... hence one would never have a Rust "cartridge" (or even a Rust "project")... one would their separate Rust code that they compile externally into a `wasm` "cartridge"... and that wasm cartridge wouldn't be related to WAT at all (other than the fact that one could potentially de-compile it).

I'm not sure we should build in the decompiling or anything like that...

So I suppose this is my proposal that if someone wants to edit WAT in the built-in editor and then compile/run that - that that would be a separate PR from the one I've just submitted... and that one would use a `script: wat` tag for that... not `script: wasm`.

**Comment by joshgoebel** [comment]
Also I think trying to mix them leads to a lot of confusing workflow questions - because the BINARY segment and CODE segments aren't guaranteed to match. Questions like:

- If a cartridge has BINARY chunk and a CODE `wat` chunk, which should take priority?  
- Does one always supersede the other?  
- If so, the binary would have to win to allow external compiled languages?
- Should the `wat` CODE always just be decompiled from the wasm then?  
- But then you'd lose the ability to comment your code, etc...

Given there are so many workflows possible here I don't see how we could ever be certain what is intended with having `wasm` be a separate thing.  I suppose you could perhaps have a `external` flag in the BINARY area to indicate that the BINARY blob was imported... but then again if we can generate it from `wat`, why would wat cartridges have a BINARY chunk saved - vs compiling real-time - which should be plenty fast enough just for startup.

So I'd suggest `wasm` always means a single thing:

- Load the BINARY chunk as `wasm` and execute

No ambiguity.  We only need to decide a "language" for the editor just to the precious `script: wasm` magic comment... @nesbox has mentioned you could encode that in the BINARY chunk perhaps, but I actually think it's a little elegant to still have a CODE chunk (with magic comments)... one use it to add a README, LICENSE text, etc...

**Comment by nesbox** [comment]
Personally, I like the idea of having two chunks `CODE` and `BINARY` with the `script: wasm` tag in the code, also it is easier for the website to parse the code chunk when adding carts.

**Comment by joshgoebel** [comment]
I think the larger question here is `wasm` vs `wat` though... I almost think just having `wat` as the language in the editor adds confusion... I almost think perhaps we should just pick Lua (for the familiar (to many) `--` comments), but then remove all the other highlighting... so that no one is inclined to start typing code there.

And of course the "new" script can have some notes to make it plain how one is to use `wasm` cartridges.

**Comment by msx80** [comment]
> The original implementation had this idea of the editor being WAT and the runtime being WASM, but I think this is flawed... the script metadata being WASM doesn't imply WAT is the source language or the language the cartridge was originally written in... now WAT is simply a textual format of WASM true, but the source language might have been C, C++, Go, Rust, or many others... so I think expecting wasm to be written in wat is already broken conceptually.

Indeed, but then again nothing in the cart implies which source language was used. In fact, there's absolutely no need to know, as WASM is self standing. From the point of view of the cartridge, there's no source language, just the wasm. In this context, wat is simply a way to represent the binary on the screen.

> I feel like if we wanted someone to use the built-in editor to write raw WAT (which doesn't seem like something many people would want), that the script would be wat... and that when you hit run the wat would be compiled to WASM and run "on the fly"... there might never be a BINARY chunk at all, least not one that was persisted.

Indeed, that was my way of thinking it before i learned about the binary chunk. It could have worked, the only issue would be for code size as WAT is pretty verbose and could easily generate huge sources.
Other languages could have been included with low level encoders (like machine code decoding etc)

> So hence a wat project might potentially make sense, but again it wouldn't have a BINARY chunk... it would just be WAT code that was compiled as needed by TIC-80 to run the cartridge. I don't see the need to store both since then you have no idea which is canonical (one has to be)... and once one is, why not just keep that? Since the source can also has comments, etc... it seems to make sense for that to be canonical.

Yes, in that context the binary chunk would be unnecessary. Assemblers would have been treated like any other language, with their interpreters.

> I'm not sure we should build in the decompiling or anything like that...

The original idea was not decompiling, it was just textual representation of the binary, which is much simpler. Every bytecode has one, and it's generally not hard to implement a decoder (which is different from, for example, and assembler, which can handle things like labels etc).

> So I suppose this is my proposal that if someone wants to edit WAT in the built-in editor and then compile/run that - that that would be a separate PR from the one I've just submitted... and that one would use a script: wat tag for that... not script: wasm.

Sounds good to me.

> Also I think trying to mix them leads to a lot of confusing workflow questions - because the BINARY segment and CODE segments aren't guaranteed to match. Questions like:

Well as i said the idea predates the binary chunk so it was just the wat code being converted on the fly to bytecode. I agree that having both is superfluous and can create confusion as you noted.

> No ambiguity. We only need to decide a "language" for the editor just to the precious script: wasm magic comment... @nesbox has mentioned you could encode that in the BINARY chunk perhaps, but I actually think it's a little elegant to still have a CODE chunk (with magic comments)... one use it to add a README, LICENSE text, etc...

I also think that the actual engine should be decided by the code chunk magic comments. So the comment would read something like:

```
-- title:  game title
-- script: wasm
-- (etc)
```
This also enables us to have all other magic comments, like "input: mouse" etc. We just have to choose a customary "comment format", could be  // or -- or anything else.
Tic80 will know which engines are binary and which are not and act accordingly.

This still leaves the fact that the code editor would show things unrelated to the binary code, it will become a "comment section" of sort, am i right ? I'm ok with that if that's the way we want to go. In this case i think we can actually drop wat altogether. Implementing it as a separated language makes no sense, nobody programs in wat directly and wasm is better served by the binary implementation.

We'll still have the option, in future, to implement something like an "hex editor" tab in tic80 that shows and edits the binary chunk.

**Comment by joshgoebel** [comment]
> The original idea was not decompiling, it was just textual representation of the binary

That's what I meant, perhaps "disassembly" was the word I'm looking for... binary -> text (assembler-ish)... I was referring to the process of changing the binary back into the more "assembly like" WAT format.

> This still leaves the fact that the code editor would show things unrelated to the binary code, it will become a "comment section" of sort, am i right?

Yes, I actually thought Markdown (with `--` comments) would be nice, but text content with just most of the editor stuff turned off isn't too bad really.  One could use this to include a README, a written LICENSE, other information, etc...

> Implementing it as a separated language makes no sense, nobody programs in wat directly 

Exactly my thinking. :-)

> to implement something like an "hex editor" tab in tic80 that shows and edits the binary chunk.

I think this would be of very limited usefulness for most people though... most working with WASM are going to be compiling from source... if you wanted to change your program you'd just recompile the original source.

Meanwhile say someone wanting to "hack" cartridges would be better off `export binary` and covert to WAT, hack, then convert back to WASM and import... that would be a much nicer experience than trying to parse the raw hex.

**Comment by joshgoebel** [comment]
I was just talking to @keithohara about the possibility of `script: wat`... (for those who REALLY want to write and run raw textual web assembly as it's own thing) and just realized that now that the WASM work is done it would be absurdly _trivial_ to add WAT (as it's own language) provided we could get the `wat2wasm` source to compile properly on all our supported platforms.

Probably no more than a 200 line wrapper around the WASM runtime that on init compiles a temporary BINARY segment and on close wipes that segment and otherwise just defers to WASM init/shutdown... everything else is just the editor trappings, syntax keywords, etc...

Now personally I don't think this is a thing a lot of people would want, but as a teaching/educational tool I could see it being interesting... provided the compiler error were surfaced well (and assuming there are decent compiler errors to be had).

**Comment by joshgoebel** [comment]
Closing via #1772. I'll open the "WAT" as it's own scripting language as a separate issue.
